python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -0,0 +1,243 @@
+    """
+    if not config:
+        config = Config()
+        config[".datastore.cls"] = "lsst.daf.butler.datastores.inMemoryDatastore.InMemoryDatastore"
+        config[".datastore.checksum"] = False  # In case of future changes
+        config[".registry.db"] = "sqlite:///:memory:"
+    Butler.makeRepo(root, config=config, **kwargs)
+    butler = Butler(root, writeable=True)
+    dimensionRecords = _makeRecords(dataIds, butler.registry.dimensions)
+    for dimension, records in dimensionRecords.items():
+        butler.registry.insertDimensionData(dimension, *records)
+    return butler
+
+
+def makeTestCollection(repo):
+    """Create a read/write Butler to a fresh collection.
+
+    Parameters
+    ----------
+    repo : `lsst.daf.butler.Butler`
+        A previously existing Butler to a repository, such as that returned by
+        `~lsst.daf.butler.Butler.makeRepo` or `makeTestRepo`.
+
+    Returns
+    -------
+    butler : `lsst.daf.butler.Butler`
+        A Butler referring to a new collection in the repository at ``root``.
+        The collection is (almost) guaranteed to be new.
+    """
+    # Create a "random" collection name
+    # Speed matters more than cryptographic guarantees
+    collection = "test" + "".join((str(i) for i in np.random.randint(0, 10, size=8)))
+    return Butler(butler=repo, run=collection)
+
+
+def _makeRecords(dataIds, universe):
+    """Create cross-linked dimension records from a collection of
+    data ID values.
+
+    Parameters
+    ----------
+    dataIds : `~collections.abc.Mapping` [`str`, `iterable`]
+        A mapping keyed by the dimensions of interest. Each value is an
+        iterable of names for that dimension (e.g., detector IDs for
+        `"detector"`).
+    universe : lsst.daf.butler.DimensionUniverse
+        Set of all known dimensions and their relationships.
+
+    Returns
+    -------
+    dataIds : `~collections.abc.Mapping` [`str`, `iterable`]
+        A mapping keyed by the dimensions of interest, giving one
+        `~lsst.daf.butler.DimensionRecord` for each input name. Related
+        dimensions (e.g., instruments and detectors) are linked arbitrarily.
+    """
+    expandedIds = {}
+    # Provide alternate keys like detector names
+    for name, values in dataIds.items():
+        expandedIds[name] = []
+        dimension = universe[name]
+        for value in values:
+            expandedValue = {}
+            for key in dimension.uniqueKeys:
+                if key.nbytes:
+                    castType = bytes
+                else:
+                    castType = key.dtype().python_type
+                try:
+                    castValue = castType(value)
+                except TypeError:
+                    castValue = castType()
+                expandedValue[key.name] = castValue
+            for key in dimension.metadata:
+                if not key.nullable:
+                    expandedValue[key.name] = key.dtype().python_type(value)
```

Comment:
Mostly to check that I'm reading this correctly (and maybe suggest that the answer be added to the docs): this just tries to populate any other required fields of the generated records by casting the data ID value for that dimension to the appropriate type for that field?

It's possible that won't work as the set of columns and types grows to realistic production complexity, but it seems like it's a reasonable approach for now.  And if/when that breaks, I'd rather just have some shared test data that provides all of the correctly-related dimension records one might want in a test (e.g. something derived from the ci_hsc output database might work), though of course I understand that this is not the ticket for such a switch.

Reply:
I'm not going for "realistic production complexity". This is specifically for unit testing, i.e. minimal working environments for a specific kind of Butler/PipelineTask functionality.

Would it help if I added something more explicit to the docs that this method is only for if you *don't care* how your test data are structured, you just want simple Butler queries to work as expected?

Reply:
Oh wait... are you saying the default config will get more complex? That would be a problem. :worried:

Reply:
Yes, default config will get more complex, but let's not worry about that now. The rest of this discussion is getting into the philosophy of how we design and test daf_butler, and we should have that (elsewhere) at some point.  Look for a community post in the next few days.